### ISSUE TITLE
empty string is passing the 'array' rule

### ISSUE THREAD

**Issue by vieks** [poppinss/indicative#41]
``` javascript
  indicative
  .validate({foo: ''}, {foo: 'array'})
  .then (() => {
    // validation is OK !
  })
```

**Comment by vieks** [comment]
It seems this is the same thing for every rules, empty string behaves as undefined.

**Comment by thetutlage** [comment]
That is true, empty values should be taken care by `required` rule, anything that is empty or undefined will skip other rules.

**Comment by vieks** [comment]
> anything that is empty or undefined will skip other rules.

This is a great behavior for undefined but not for empty string.
Indeed, undefined mean: '**no data** => field is not set by user'
whereas empty string mean: "**data exists** => field is set by user with empty string as value"

So how to handle with this empty string behavior the following validation case:
The input field named 'foo' is optional and must be an array if settled ?

My prior solution wasn't good:

``` javascript
.validate({foo: ''}, {foo: 'array'})
```

because the user send me the value empty string as 'foo' input field.

In fact 'indicative' have the good behavior for all other types:
`.validate({foo: null}, {foo: 'array'})` => fails ! ok
`.validate({foo: 0}, {foo: 'array'})` => fails ! ok
`.validate({foo: undefined}, {foo: 'array'})` => passes ! ok
`.validate({}, {foo: 'array'})` => passes ! ok (it's the same as the previous line, treated as undefined it's great)
`.validate({foo: false}, {foo: 'array'})` => fails ! ok
`.validate({foo: ''}, {foo: 'array'})` => passes ! NO this breaks the validation indeed user submits the wrong value: an empty string and we need an array.

I think to know why you made this design choice: With **old and classical html forms** an optional input field:

``` html
<input type="text" name="foo" />
```

sending with no values will give a x-www-form-urlencoded structure: foo=&

on http server (express)

``` javascript
app.use(bodyParser.urlencoded({ extended: false }));

//    parsing output
const data = { foo: '' }
```

...so later when validation comes:

``` javascript
 indicative
  .validate(data, {foo: 'integer'})
  .then (() => {
    // validation is OK ! GREAT !!!
  })
```

And indeed for that use case this is the correct behavior and that's exactly what indicative implements !
But for more modern applications this approach sucks. Data from user input and server output are always JSON.

So for example for my use case with an api:

filters is optional
`GET 127.0.0.1:8080/api/v1/users?filters={ "limit": 2, "select": ["id","name"] }`

``` javascript
//    after app parsing
const query = { limit: 2, select: ['id', 'name'] }


// validation
 indicative
  .validate(query, { limit: 'number|above:-1|under:51', select: 'array'})
  .then (() => {
    // validation is OK ! GREAT !!!
  })
```

with this:
select is an empty array
`GET 127.0.0.1:8080/api/v1/users?filters={ "limit": 2, "select": [] }`
or
select is undefined
`GET 127.0.0.1:8080/api/v1/users?filters={ "limit": 2 }`

validation passes again for both, great !

but:
`GET 127.0.0.1:8080/api/v1/users?filters={ "limit": 2, "select": "" }`
validation passes again despite the empty string which not respect my optional array rule !
And that's really bad, because I need the select user input to be an **array** or **nothing** (undefined) but not a string type (and empty furthermore)...

Indeed later, at database retrieving step, I do something like that:

``` javascript
getAll: function(limit=30, offset=0, select=['name']){
    ...fetch from db
}
```

But with the validation accepting an empty string It breaks my query requiring an array or undefined using in that case the default parameter, but never an empty string which will overpass my default param.

Yes I can use workaround solutions like sanitizing all user inputs from `''` to `undefined` with the built-in`delete` after server parsing but it's quite ugly and can be high recursive in the only purpose to satisfy the validation step.

Or do a

``` javascript
//    select == ''
getAll: function(limit=30, offset=0, select){
     select = select || ['name'];
    ...fetch from db
}
```

Since I use ES6 default parameters syntax doing that kind of patch for all my code handling user inputs is also quite ugly.

For both cases it's really cumbersome you will agree especially when we know how much you are attached to produce beautiful code :)

A solution for indicative will be to allow to choose a string mode to handle them with a `indicative.setMode('string strict')` to choose whether an empty string is considered as a value or not (treated as undefined). Or of course whatever solution you want and looks beautiful.

Otherwise I will haven't choice and must implement validation on my own, pretty bad since indicative engine have no flaws apart from that case and is really handy, powerful and beautiful. All others js validations tools are ugly and bloated: joi, checkit, validate.js...

**Comment by thetutlage** [comment]
Let me run tests with different behaviors, working on it

**Comment by vieks** [comment]
nice, thank you  :+1:

**Comment by thetutlage** [comment]
Released as `2.0.2` on npm
